src/Testable/Navigation.elm

```diff
@@ -0,0 +1,156 @@
+    List.Extra.getAt index history
+        |> Maybe.withDefault (getLocation "")
+
+
+update : Msg -> History -> ( History, ReturnMsg )
+update msg ( index, history ) =
+    let
+        location =
+            currentLocation ( index, history )
+    in
+    case msg of
+        Jump n ->
+            let
+                nextIndex =
+                    min (max (index + n) 0) (List.length history - 1)
+
+                nextLocation =
+                    List.Extra.getAt nextIndex history
+                        |> Maybe.withDefault (getLocation "")
+            in
+            ( ( nextIndex, history ), TriggerLocationMsg nextLocation )
+
+        New url ->
+            let
+                nextLocation =
+                    setLocation url location
+
+                modifiedHistory =
+                    List.take (index + 1) history ++ [ nextLocation ]
+            in
+            ( ( index + 1, modifiedHistory ), ReturnLocation nextLocation )
+
+        Modify url ->
+            let
+                nextLocation =
+                    setLocation url location
+
+                modifiedHistory =
+                    history
+                        |> List.Extra.updateIfIndex ((==) index) (always nextLocation)
+
+                effect =
+                    if nextLocation == location then
+                        NoOp
+                    else
+                        ReturnLocation nextLocation
+            in
+            ( ( index, modifiedHistory ), effect )
+
+        Visit url ->
+            let
+                ( modifiedHistory, _ ) =
+                    update (New url) ( index, history )
+            in
+            ( modifiedHistory, TriggerLocationMsg (currentLocation modifiedHistory) )
+
+
+init : History
+init =
+    ( 0, [ initialLocation ] )
+
+
+initialLocation : Location
+initialLocation =
+    getLocation "https://elm.testable/"
+
+
+getLocation : String -> Navigation.Location
+getLocation href =
+    let
+        parser =
+            find All (regex "(.*?:)//(.*?):?(\\d+)?(/[^\\?^#]*)(\\?[^#]*)?(#.*)?") href
+
+        matchAt index =
+            List.head parser
+                |> Maybe.andThen
+                    (.submatches
+                        >> List.Extra.getAt (index - 1)
+                        >> Maybe.withDefault Nothing
+                    )
+
+        matchOrEmptyAt index =
+            Maybe.withDefault "" (matchAt index)
+
+        protocol =
+            matchOrEmptyAt 1
+
+        host =
+            matchOrEmptyAt 2
+
+        pathname =
+            Maybe.withDefault "/" (matchAt 4)
+    in
+    { href = href
+    , host = host
+    , hostname = host
+    , protocol = protocol
+    , origin = protocol ++ "//" ++ host
+    , port_ = matchOrEmptyAt 3
+    , pathname =
+        if String.isEmpty pathname then
+            "/"
+        else
+            pathname
+    , search = matchOrEmptyAt 5
+    , hash = matchOrEmptyAt 6
+    , username = ""
+    , password = ""
+    }
```

Comment:
Is there not another package that does this correctly?  If not, we should look at making/publishing that package (not necessary for this PR, though).